qiskit/providers/ibmq/api/clients/websocket.py

```diff
@@ -161,75 +180,119 @@ def get_job_status(
+                            warnings.filterwarnings("ignore", category=DeprecationWarning)
+                            if timeout:
+                                response_raw = yield from asyncio.wait_for(
+                                    websocket.recv(), timeout=timeout)
+
+                                # Decrease the timeout, with a 5-second grace period.
+                                elapsed_time = time.time() - start_time
+                                timeout = max(5, int(original_timeout - elapsed_time))
+                            else:
+                                response_raw = yield from websocket.recv()
+                        logger.debug('Received message from websocket: %s',
+                                     response_raw)
+
+                        # Successfully received a message, reset retry counter.
+                        current_retry_attempt = 0
+
+                        response = WebsocketMessage.from_bytes(response_raw)
+                        last_status = response.data
+
+                        job_status = response.data.get('status')
+                        if (job_status and
+                                ApiJobStatus(job_status) in API_JOB_FINAL_STATES):
+                            break
+
+                    except futures.TimeoutError:
+                        # Timeout during our wait.
+                        raise WebsocketTimeoutError('Timeout reached') from None
+                    except ConnectionClosed as ex:
+                        # From the API:
+                        # 4001: closed due to an internal errors
+                        # 4002: closed on purpose (no more updates to send)
+                        # 4003: closed due to job not found.
+                        message = 'Unexpected error'
+                        if ex.code == 4001:
+                            message = 'Internal server error'
+                        elif ex.code == 4002:
+                            break
+                        elif ex.code == 4003:
+                            attempt_retry = False  # No point in retrying.
+                            message = 'Job id not found'
+
+                        raise WebsocketError('Connection with websocket closed '
+                                             'unexpectedly: {}(status_code={})'
+                                             .format(message, ex.code)) from ex
+
+            # Errors unrelated to websocket connection, not worth retrying.
+            except (WebsocketIBMQProtocolError, WebsocketTimeoutError) as ex:
+                raise ex
+
+            # Errors related to websocket connection, attempt retrying.
+            except (WebsocketError, WebsocketAuthenticationError) as ex:
+                logger.warning('%s', ex)
+
+                current_retry_attempt = current_retry_attempt + 1
+                if current_retry_attempt > retries:
+                    break  # Attempted maximum number of retries already.
```

Comment:
It's probably worth keeping `WebsocketAuthenticationError` instead of masking it as `WebsocketError`, just to know that the error occurred during authentication. Or at least do a `from ex` so the previous error is preserved.